docs/stdlibs.md

```diff
@@ -0,0 +1,187 @@
+3. Its native Go performance significantly outperforms the Gno counterpart by
+   several orders of magnitude, and it is used in crucial code or hot paths in
+   many programs.\
+   For example: `sha256.Sum256`.
+
+The takeaway here is that native bindings are a special feature which can be
+useful to overcome pure Gno limitations, but it is not a substitute for writing
+standard libaries in Gno.
+
+There are three components to a natively bound function in Gno:
+
+1. The Gno function definition, which must be a top-level function with no body
+   (and no brackets).\
+   For example: `crypto/sha256/sha256.gno`.
+2. The Go function declaration, which must be a top-level function with the same
+   name and signature.\
+   For example: `crypto/sha256/sha256.go`.
+3. When the two above are present and valid, the native binding can be created
+   by executing the code generator: either execute `go generate` from the
+   `stdlibs` directory, or run `make generate` from the `gnovm` directory.\
+   This generates the `native.go` file available in the `stdlibs` directory,
+   which provides the binding itself to then be used by the GnoVM.
+
+The code generator in question is available in the `misc/genstd` directory.
+There are some quirks and features that must be kept in mind when writing native
+bindings, which are the following:
+
+- Unexported functions (for instance, `func sum256(b []byte)`) must have their
+  Go counterpart prefixed with `X_` in order to make the functions exported (ie.
+  `func X_sum256(b []byte)`).
+- The Go function declaration may specify as the first argument
+  `m *gno.Machine`, where `gno` is an import for
+  `github.com/gnolang/gno/gnovm/pkg/gnolang`. This gives the function access to
+  the Virtual Machine state, and is used by functions like `std.AssertOriginCall()`.
+- The Go function may change the type of any parameter or result to
+  `gno.TypedValue` (where `gno` is an import for the above import path). This
+  means that the `native.go` generated code will not attempt to automatically
+  convert the Gno value into the Go value, and can be useful for unsupported
+  conversions like interface values.
+- A small set of named types are "linked" between their Gno version and Go
+  counterpart. For instance, `std.Address` in Gno is
+  `(".../tm2/pkg/crypto").Bech32Address` in Go. A list of these can be found in
+  `misc/genstd/mapping.go`.
+- Not all type literals are currently supported when converting from their Gno
+  version to their Go counterpart. Notable omissions at the time of writing
+  include struct and map literals. If you intend to use these, modify the code
+  generator to support them.
+- The code generator does not inspect any imported packages from the Go native code;
+  as a consequence, either the import contains an identifier (ie.
+  `import gno "github.com/gnolang/gno/gnovm/pkg/gnolang"`), or the last element
+  in the import path is assumed to be the identifier used in the file. Note that
```

Comment:
i'm confused by this

Reply:
I'll try to make this pragmatic with an example.

The tl;dr is:

```go
// foo/bar/doc.go
package baz

func X() {}
```

```go
// main.go
package main

import "foo/bar"

func main() { baz.X() }
```

The following two are normal in Go code; in main.go the identifier for the import of `foo/bar` will be what the `package` clause is inside of `foo/bar/*.go`.

To avoid resolving package dependencies in the code generator, I simplified this, so if the generator inspects `main.go` it will expect `import "foo/bar"` to have the identifier `bar`.

Like with the example above, this is not always correct. In that case, the author can fix that by making main.go the following:


```go
// main.go
package main

import baz "foo/bar"

func main() { baz.X() }
```

ie. adding an explicit identifier.

Reply:
hows the fish?

Reply:
s/The code generator does not inspect any imported packages from the Go native code;/
The code generator does not inspect the imported native Go packages to get the internally declared package name;/